src/utils/utils.ts

```diff
@@ -70,6 +70,39 @@ async function* arrayToAsyncIterable<T>(array: T[]): AsyncIterable<T> {
   }
 }
 
+/*
+This function:
+1. Keeps regular spaces, only ' ', as they are.
+2. Converts \n \r \t to escaped versions, \\n \\r and \\t.
+3. Converts other control characters to their Unicode escape sequences.
+ */
+
+function encodeNonPrintable(input) {
+  // Ensure the input is a string
+  const str = String(input);
+  // eslint-disable-next-line
```

Comment:
1. Can't you specify `input: string` if you want it to be a string? Why cast it here with `String()`?
2. never use `eslint-disable-next-line` unless you absolutely have to, and if you do you *must* include the reasoning with a comment. And you must specify the exact rule you are ignoring line ` eslint-disable-next-line no-control-regex`. Generally we want a very good reason for doing this.

Reply:
the Docbloc has been cleaned up, the type signature now takes a string, and a comment has been added for reasoning as to why `eslint-disable-next-line no-control-regex` is used
```ts
/**
 * This function:
 * 1. Keeps regular spaces, only ' ', as they are.
 * 2. Converts \n \r \t to escaped versions, \\n \\r and \\t.
 * 3. Converts other control characters to their Unicode escape sequences.
 */
function encodeNonPrintable(str: string) {
  // We want to actually match control codes here!
  // eslint-disable-next-line no-control-regex
  return str.replace(/[\x00-\x1F\x7F-\x9F]/g, (char) => {
    switch (char) {
      case ' ':
        return char; // Preserve regular space
      case '\n':
        return '\\n'; // Encode newline
      case '\r':
        return '\\r'; // Encode carriage return
      case '\t':
        return '\\t'; // Encode tab
      case '\v':
        return '\\v'; // Encode tab
      case '\f':
        return '\\f'; // Encode tab
      // Add cases for other whitespace characters if needed
      default:
        // Return the Unicode escape sequence for control characters
        return `\\u${char.charCodeAt(0).toString(16).padStart(4, '0')}`;
    }
  });
}
```